src/python/pants/backend/go/module.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import textwrap
+from dataclasses import dataclass
+from typing import List, Optional, Tuple
+
+import ijson  # type: ignore
+
+from pants.backend.go.distribution import GoLangDistribution
+from pants.backend.go.target_types import GoModuleSources
+from pants.build_graph.address import Address
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Addresses
+from pants.engine.console import Console
+from pants.engine.fs import (
+    CreateDigest,
+    Digest,
+    DigestContents,
+    FileContent,
+    MergeDigests,
+    RemovePrefix,
+    Snapshot,
+    Workspace,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get
+from pants.engine.platform import Platform
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import UnexpandedTargets
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+@dataclass(frozen=True)
+class ModuleDescriptor:
+    import_path: str
+    module_path: str
+    module_version: str
+
+
+@dataclass(frozen=True)
+class ResolvedGoModule:
+    import_path: str
+    minimum_go_version: Optional[str]
+    modules: FrozenOrderedSet[ModuleDescriptor]
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class ResolveGoModuleRequest:
+    address: Address
+
+
+# Perform a minimal parsing of go.mod for the `module` and `go` directives. Full resolution of go.mod is left to
+# the go toolchain. This could also probably be replaced by a go shim to make use of:
+# https://pkg.go.dev/golang.org/x/mod/modfile
+def basic_parse_go_mod(raw_text: bytes) -> Tuple[Optional[str], Optional[str]]:
+    module_path = None
+    minimum_go_version = None
+    for line in raw_text.decode("utf-8").splitlines():
+        parts = line.strip().split()
+        if len(parts) >= 2:
+            if parts[0] == "module":
+                if module_path is not None:
+                    raise ValueError("Multiple `module` directives found in go.mod file.")
+                module_path = parts[1]
+            elif parts[0] == "go":
+                if minimum_go_version is not None:
+                    raise ValueError("Multiple `go` directives found in go.mod file.")
```

Comment:
Ik you're still iterating, but it would be good to have better error messages here like saying which file is the problem.

If you want to punt, you could add a TODO.

Reply:
Will add TODOs.